### ISSUE TITLE
The conditional check 'tor_v4ips == []' failed

### ISSUE THREAD

**Issue by lunarthegrey** [nusenu/ansible-relayor#245]
**Describe the bug**

When running a playbook with the role, I get an error when running this task:

```
TASK [nusenu.relayor : Use a single private IPv4 address if we have no public IPv4 address] 
```

The error states:

```
"msg": "The conditional check 'tor_v4ips == []' failed. The error was: An unhandled exception occurred while templating '{{ tor_available_public_ipv4s[0:tor_maxPublicIPs] }}'
```

Debug information is at the bottom.

**To Reproduce**

Run a playbook utilizing nusenu.relayor

**Expected behavior**

The playbook runs successfully.

**Version information (please include the following information):**
- ansible version:

```
ansible [core 2.16.4]
  config file = /Users/user/.ansible.cfg
  configured module search path = ['/Users/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/homebrew/Cellar/ansible/9.3.0/libexec/lib/python3.12/site-packages/ansible
  ansible collection location = /Users/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /opt/homebrew/bin/ansible
  python version = 3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)] (/opt/homebrew/Cellar/ansible/9.3.0/libexec/bin/python)
  jinja version = 3.1.3
  libyaml = True
```

- ansible-relayor version (please try to reproduce the bug against the git master branch): `v23.1.0`

**Playbook information**

```
- name: Run the nusenu.relayor role and include custom vars
  remote_user: root
  gather_facts: yes
  hosts: '{{ target }}'
  vars_files:
    - vars/main.yml
  roles:
    - nusenu.relayor
```

**OS information**

Control: macOS 14.4
Target: Debian 12

**Debug information**

```
TASK [nusenu.relayor : Use a single private IPv4 address if we have no public IPv4 address] ***************************************************************
task path: /Users/user/.ansible/roles/nusenu.relayor/tasks/ip-list.yml:3
redirecting (type: filter) ansible.builtin.ipv4 to ansible.netcommon.ipv4
[DEPRECATION WARNING]: Use 'ansible.utils.ipv4' module instead. This feature will be removed from ansible.netcommon in a release after 2024-01-01. 
Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
redirecting (type: filter) ansible.netcommon.ipv4 to ansible.utils.ipv4
fatal: [target-host]: FAILED! => {
    "msg": "The conditional check 'tor_v4ips == []' failed. The error was: An unhandled exception occurred while templating '{{ tor_available_public_ipv4s[0:tor_maxPublicIPs] }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while templating '{{ ansible_all_ipv4_addresses| ipv4('address') | ipv4('public') }}'. Error was a <class 'AttributeError'>, original message: 'IPAddress' object has no attribute 'is_private'\n\nThe error appears to be in '/Users/user/.ansible/roles/nusenu.relayor/tasks/ip-list.yml': line 3, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Use a single private IPv4 address if we have no public IPv4 address\n  ^ here\n"
}
```

**Comment by nusenu** [comment]
The most likely reason is that you are missing python-netaddr on your control node as mentioned in the requirements.

* is python-netaddr installed on your macOS as mentioned in the requirements? https://github.com/nusenu/ansible-relayor?tab=readme-ov-file#requirements
* please paste the content of your vars file into the issue
* in general please test against the latest relayor release or master
* you appear to be using remote root logins which is not supported by this role (README: "do not run this role with become: yes")

**Comment by lunarthegrey** [comment]
Thanks @nusenu.

netaddr appears to be installed even though its not available in homebrew anymore:

```
user@machine ~ % ls -lah /opt/homebrew/bin/netaddr
-rwxr-xr-x  1 user  admin   238B Nov  6  2022 /opt/homebrew/bin/netaddr
```

I even tried installing it with `pipx` and it didn't change anything:

```
user@machine ~ % pipx install netaddr       
⚠️  Note: netaddr was already on your PATH at /opt/homebrew/bin/netaddr
  installed package netaddr 1.2.1, installed using Python 3.12.2
  These apps are now globally available
    - netaddr
⚠️  Note: '/Users/user/.local/bin' is not on your PATH environment variable. These apps will not be globally accessible until your PATH is updated. Run `pipx ensurepath` to
    automatically add it, or manually modify your PATH in your shell's config file (i.e. ~/.bashrc).
done! ✨ 🌟 ✨
```

```
user@machine ~ % pipx ensurepath
Success! Added /Users/user/.local/bin to the PATH environment variable.

Consider adding shell completions for pipx. Run 'pipx completions' for instructions.

You will need to open a new terminal or re-login for the PATH changes to take effect.

Otherwise pipx is ready to go! ✨ 🌟 ✨
```

Reopened new terminal, reran playbook and it produced the same error.

> in general please test against the latest relayor release or master

I am on the latest version in Ansible Galaxy.

> you appear to be using remote root logins which is not supported by this role (README: "do not run this role with become: yes")

I run this with root because of my pre and post tasks which do extra things that require root, although I do not believe this is the cause of the error.

> please paste the content of your vars file into the issue

Here is my vars.yaml:

```
---

# Non-default vars
tor_ContactInfo: "redacted"
tor_nicknamefile: files/nicknames.csv
tor_AbuseEmailAddress: "redacted"

tor_user: _tor
tor_packages: tor
tor_package_state: latest
tor_binary: tor

tor_config: {}
tor_prom_labels: {}
# example:
# tor_prom_labels:
#   label1: "value1"
#   label2: "value2"

tor_gen_ciiss_proof_files: False
tor_ciiss_proof_folder: '~/.tor'

# The tor_ports dictionary defines how many instances are created per
# available IP address and what ports are used.
# min entries: 1
# max entries: 2
tor_ports:
  - orport: 443
    dirport: 80

# Limit the amount of used IP addresses
# by default to 1, this limits the amount
# of generated Tor instances to 2 per host.
# Specifying tor_ips overwrites this limitation.
tor_maxPublicIPs: 1

# We only use a single private IP by default since
# we can not reliably tell whether we are NATed
# to more than a single public IP address
tor_maxPrivateIPs: 0

# we are a non-exit by default
# setting this to true will enable reduced exit policy
# and allow custom exit policies via tor_ExitPolicy
# Logic is implemented in templates/torrc
tor_ExitRelay: True

# deploy default tor-exit-notice.html if we are an exit
tor_ExitNoticePage: True

# automatically detect IPv6 addresses and enable IPv6 ORPort
tor_IPv6: True

# enable IPv6 exiting
# the following var only matters if
# tor_ExitRelay and tor_IPv6 is True
tor_IPv6Exit: True

# on FreeBSD we increase the following kernel
# settings if they are lower than that.
# minimal kern.ipc.somaxconn value
tor_freebsd_somaxconn: 1024
# minimal kern.ipc.nmbclusters value
tor_freebsd_nmbclusters: 30000

# Attention: we will run chown tor_user on these folders
tor_DataDir: /var/lib/tor
tor_PidDir: /var/run/tor-instances

# Create a backup of the torrc file on the tor server whenever the file is changed.
# You might want to disable this if you use other means of backups or version
# control, for example etckeeper.
tor_backup_torrc: yes

# specify tor's loglevel
tor_LogLevel: notice

# Where do you want to store key material on the
# ansible control machine?
tor_offline_masterkey_dir: ~/.tor/offlinemasterkeys

# How long should online keys/certs be valid before expiring?
tor_signingkeylifetime_days: 90

# # this var is used for apt sources.list entries
# tor_distribution_release: "{{ ansible_lsb.codename }}"

# # filter potentially malicious input from relay-provided IPv4/v6 addresses
# # https://docs.ansible.com/ansible/playbooks_filters_ipaddr.html
# tor_available_public_ipv4s: "{{ ansible_all_ipv4_addresses| ipv4('address') | ipv4('public') }}"
# tor_v4ips: "{{ tor_available_public_ipv4s[0:tor_maxPublicIPs] }}"
# tor_ipv4_count: "{{ tor_v4ips | length|int }}"

# # we can not use more IPv6 IPs than we have IPv4 IPs so we truncate (but fewer is ok)
# tor_available_public_ipv6s: "{{ ansible_all_ipv6_addresses|ipv6('public')|ipv6('address') }}"

tor_apt_update_cache: yes

# This var enables autoconfiguration for OutboundBindAddressExit
tor_dedicatedExitIP: False

# the following line is commented out and handled in ip-list.yml until
# https://github.com/ansible/ansible/issues/14829
# gets fixed
#tor_v6ips: "{{ tor_v6tmp[0:tor_ipv4_count|int]|ipv6('address') }}"

tor_RunAsDaemon: 1

tor_enableControlSocket: False

# setting tor_alpha to True
# will install tor alpha repos.
# This is only supported for Debian/Ubuntu
tor_alpha: False
tor_nightly_builds: False

# known problem: requires manual adjustment with every major tor release
tor_alpha_version: experimental

# MetricsPort variables
tor_enableMetricsPort: False
tor_MetricsPort_offset: 33300

# prometheus (tor MetricsPort) related vars
# On which host should we generate the prometheus scrape config? (delegate_to)
# tor_prometheus_host: 127.0.0.1
# # file path to the scrape config
# tor_prometheus_scrape_file: "~/.tor/tor-prometheus-scrape-configs"
# tor_blackbox_exporter_host: localhost:9115
# tor_gen_blackbox_scrape_config: False
# tor_blackbox_exporter_scheme: http
# tor_blackbox_exporter_password: "{{ lookup('password', '~/.tor/blackbox_exporter_password') }}"

# tor_prometheus_scrape_password_folder: "~/.tor/prometheus/scrape-passwords/"
# tor_prometheus_scrape_port: 443
# tor_gen_metricsport_htpasswd: False
# tor_metricsport_htpasswd_file: "/etc/nginx/tor_metricsport_htpasswd"
# tor_metricsport_htpasswd_file_owner: "www-data"

# Path to the tor-exit-notice HTML file
tor_exit_notice_file: tor-exit-notice.html

# List of blacklisted DNS resolvers
#
# Google DNS IPs:
#  8.8.8.8
#  8.8.4.4
#  2001:4860:4860::8888
#  2001:4860:4860::8844
#  taken from: https://developers.google.com/speed/public-dns/docs/using#google_public_dns_ip_addresses
#
# OpenDNS IPs:
#  208.67.222.123
#  208.67.220.123
#  208.67.222.222
#  208.67.220.220
#  2620:119:35::35
#  2620:119:53::53
#  taken from: https://www.opendns.com/setupguide/
#  taken from: https://www.opendns.com/about/innovations/ipv6/
#
# CloudFlare DNS IPs:
#  1.1.1.1
#  1.0.0.1
#  2606:4700:4700::1111
#  2606:4700:4700::1001
#  taken from: https://1.1.1.1/
#
# Quad9 DNS IPs:
#  9.9.9.9
#  9.9.9.10
#  149.112.112.10
#  149.112.112.112
#  2620:fe::fe
#  2620:fe::9
#  2620:fe::10
#  2620:fe::fe:10
#  taken from: https://www.quad9.com/faq/
#
# Level 3
#  4.2.2.1-6
#
# Potential future candidates:
# ----------------------------
#
# ISC.org: 204.152.184.76
#
# Background:
#  https://nymity.ch/dns-traffic-correlation/.
#  https://medium.com/@nusenu/who-controls-tors-dns-traffic-a74a7632e8ca

tor_dnsresolver_blacklist:
  - 8\.8\.8\.8
  - 8\.8\.4\.4
  - 2001:4860:4860:.*:8888
  - 2001:4860:4860:.*:8844
  - 208\.67\.222\.123
  - 208\.67\.220\.123
  - 208\.67\.222\.222
  - 208\.67\.220\.220
  - 2620:119:35:.*:35
  - 2620:119:53:.*:53
  - 1\.1\.1\.1
  - 1\.0\.0\.1
  - 2606:4700:4700:.*:1111
  - 2606:4700:4700:.*:1001
  - 9\.9\.9\.9
  - 9\.9\.9\.10
  - 149\.112\.112\.10
  - 149\.112\.112\.112
  - 2620:fe:.*:fe
  - 2620:fe:.*:9
  - 2620:fe:.*:10
  - 2620:fe:.*:fe:10
  - 4\.2\.2\.[1-6]

# Construction of grep command to
# match blacklisted DNS resolvers
tor_resolvconf_path: /etc/resolv.conf
tor_dns_resolver_blacklist_pattern: "{% for IP in tor_dnsresolver_blacklist %}{{ '-e \"^nameserver ' + IP + '[[:space:]]*$\" ' }}{% endfor %}"
tor_grep_blacklisted_dnsresolvers: "{{ [ 'grep -qE', tor_dns_resolver_blacklist_pattern, tor_resolvconf_path ] | join(' ') }}"

# Default exit policy if the user wants to be an exit but doesn't specify one
# taken from https://trac.torproject.org/projects/tor/wiki/doc/ReducedExitPolicy
tor_ExitPolicy:
#  - accept *:20-22 # disabled due to abuse
  - accept *:43
  - accept *:53
  - accept *:79-81
  - accept *:194
  - accept *:220
  - accept *:389
  - accept *:443
#  - accept *:465 # disabled due to abuse
  - accept *:531
  - accept *:543-544
  - accept *:554
  - accept *:563
#  - accept *:587 # disabled due to abuse
  - accept *:636
  - accept *:706
  - accept *:853
  - accept *:873
  - accept *:902-904
  - accept *:981
  - accept *:989-995
  - accept *:1194
  - accept *:1220
  - accept *:1293
  - accept *:1500
  - accept *:1533
  - accept *:1677
  - accept *:1723
  - accept *:1755
  - accept *:1863
  - accept *:2082
  - accept *:2083
  - accept *:2086-2087
  - accept *:2095-2096
  - accept *:2102-2104
  - accept *:3128
  - accept *:3690
  - accept *:4321
  - accept *:4643
  - accept *:5050
  - accept *:5190
  - accept *:5222-5223
  - accept *:5228
  - accept *:5900
  - accept *:6660-6669
  - accept *:6679
  - accept *:6697
  - accept *:8000
  - accept *:8008
  - accept *:8074
  - accept *:8080
  - accept *:8082
  - accept *:8087-8088
  - accept *:8332-8333
  - accept *:8443
  - accept *:8888
  - accept *:9418
  - accept *:9999
  - accept *:10000
  - accept *:11371
  - accept *:19294
  - accept *:19638
  - accept *:50002
  - accept *:64738
  - reject *:*
```

**Comment by nusenu** [comment]
btw: 
this is unrelated to your issue but 
`tor_maxPrivateIPs: 0`

is probably not what you want to set.
Maybe I should add that var to the documentation and explain it better.

**Comment by nusenu** [comment]
upstream fixes in ansible.utils (unreleased) for this issue:
https://github.com/ansible-collections/ansible.utils/issues/331
https://github.com/ansible-collections/ansible.utils/pull/338

**Comment by lunarthegrey** [comment]
Thanks @nusenu, I've been debugging for hours now and can't figure it out.

- I installed netaddr 0.10.1 and also tried 0.10.0. I completely uninstalled ansible & netaddr homebrew packages to prevent conflicts.
- I installed ansible and netaddr 0.10.x via pipx.
- I am using the git version of ansible-relayor and uninstalled the galaxy version.

Ansible version details:

```
lunar@moon ~ % ansible --version
ansible [core 2.16.6]
  config file = /Users/lunar/.ansible.cfg
  configured module search path = ['/Users/lunar/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/lunar/Library/pipx/venvs/ansible/lib/python3.12/site-packages/ansible
  ansible collection location = /Users/lunar/.ansible/collections:/usr/share/ansible/collections
  executable location = /Users/lunar/.local/bin/ansible
  python version = 3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 15.0.0 (clang-1500.3.9.4)] (/Users/lunar/Library/pipx/venvs/ansible/bin/python)
  jinja version = 3.1.3
  libyaml = True
```

Python site-packages:

```
lunar@moon ~ % ls -lah '/Users/lunar/Library/pipx/venvs/ansible/lib/python3.12/site-packages/'       
total 432
drwxr-xr-x@ 26 lunar  staff   832B Apr 20 03:25 .
drwxr-xr-x@  3 lunar  staff    96B Apr 20 03:25 ..
drwxr-xr-x@  9 lunar  staff   288B Apr 20 03:25 Jinja2-3.1.3.dist-info
drwxr-xr-x@  8 lunar  staff   256B Apr 20 03:25 MarkupSafe-2.1.5.dist-info
drwxr-xr-x@  8 lunar  staff   256B Apr 20 03:25 PyYAML-6.0.1.dist-info
-rwxr-xr-x@  1 lunar  staff   211K Apr 20 03:25 _cffi_backend.cpython-312-darwin.so
drwxr-xr-x@  4 lunar  staff   128B Apr 20 03:25 _yaml
drwxr-xr-x@ 26 lunar  staff   832B Apr 20 03:25 ansible
drwxr-xr-x@ 10 lunar  staff   320B Apr 20 03:25 ansible-9.4.0.dist-info
drwxr-xr-x@ 52 lunar  staff   1.6K Apr 20 03:25 ansible_collections
drwxr-xr-x@  9 lunar  staff   288B Apr 20 03:25 ansible_core-2.16.6.dist-info
drwxr-xr-x@  8 lunar  staff   256B Apr 20 03:25 ansible_test
drwxr-xr-x@ 25 lunar  staff   800B Apr 20 03:25 cffi
drwxr-xr-x@  9 lunar  staff   288B Apr 20 03:25 cffi-1.16.0.dist-info
drwxr-xr-x@ 11 lunar  staff   352B Apr 20 03:25 cryptography
drwxr-xr-x@ 10 lunar  staff   320B Apr 20 03:25 cryptography-42.0.5.dist-info
drwxr-xr-x@ 29 lunar  staff   928B Apr 20 03:25 jinja2
drwxr-xr-x@  9 lunar  staff   288B Apr 20 03:25 markupsafe
drwxr-xr-x@ 18 lunar  staff   576B Apr 20 03:25 packaging
drwxr-xr-x@  9 lunar  staff   288B Apr 20 03:25 packaging-24.0.dist-info
-rw-r--r--@  1 lunar  staff    62B Apr 20 03:25 pipx_shared.pth
drwxr-xr-x@ 16 lunar  staff   512B Apr 20 03:25 pycparser
drwxr-xr-x@  8 lunar  staff   256B Apr 20 03:25 pycparser-2.22.dist-info
drwxr-xr-x@ 15 lunar  staff   480B Apr 20 03:25 resolvelib
drwxr-xr-x@  8 lunar  staff   256B Apr 20 03:25 resolvelib-1.0.1.dist-info
drwxr-xr-x@ 21 lunar  staff   672B Apr 20 03:25 yaml
lunar@moon ~ % ls -lah '/Users/lunar/Library/pipx/venvs/netaddr/lib/python3.12/site-packages' 
total 8
drwxr-xr-x@  5 lunar  staff   160B Apr 20 03:32 .
drwxr-xr-x@  3 lunar  staff    96B Apr 20 03:32 ..
drwxr-xr-x@ 12 lunar  staff   384B Apr 20 03:32 netaddr
drwxr-xr-x@ 11 lunar  staff   352B Apr 20 03:32 netaddr-0.10.0.dist-info
-rw-r--r--@  1 lunar  staff    62B Apr 20 03:32 pipx_shared.pth
```

I still get the following error:

```
TASK [ansible-relayor : Use a single private IPv4 address if we have no public IPv4 address] ***
fatal: [exit-relay1-tor-ce]: FAILED! => {"msg": "The conditional check 'tor_v4ips == []' failed. The error was: An unhandled exception occurred while templating '{{ tor_available_public_ipv4s[0:tor_maxPublicIPs] }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while templating '{{ ansible_all_ipv4_addresses |  ansible.utils.ipv4('address') | ansible.utils.ipaddr('public') }}'. Error was a <class 'ansible.errors.AnsibleFilterError'>, original message: The ipv4 filter requires python's netaddr be installed on the ansible controller\n\nThe error appears to be in '/Users/lunar/Documents/ansible/ansible-relayor/tasks/ip-list.yml': line 3, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Use a single private IPv4 address if we have no public IPv4 address\n  ^ here\n"}
```

Any ideas? Thanks for your help.

**Comment by lunarthegrey** [comment]
Wow I figured it out, what a headache netaddr has caused!

To fix all of this I had to follow the below steps on macOS:

- Uninstall ansible via homebrew.
- Install ansible via pipx: `pipx install --include-deps ansible`
- Install old netaddr with pipx: `pipx install netaddr==0.10.1`
- Inject old netaddr into ansible venv: `pipx inject ansible netaddr==0.10.1`

Then I ran the playbook and I didn't hit the error again. Thanks for your help @nusenu!
